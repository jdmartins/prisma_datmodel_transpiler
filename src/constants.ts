export const TYPE_DEFINITION = 'TYPE_DEFINITION'
export const OPEN_BRACES = 'OPEN_BRACES'
export const CLOSE_BRACES = 'CLOSE_BRACES'
export const EQUAL = 'EQUAL'
export const TYPE = 'TYPE'
export const ID = 'ID'
export const DATETIME = 'DATETIME'
export const STRING = 'STRING'
export const BOOLEAN = 'BOOLEAN'
export const FLOAT = 'FLOAT'
export const INT = 'INT'
export const ENUM_DEFINITION = 'ENUM_DEFINITION'
export const IS_REQUIRED = 'IS_REQUIRED'
export const OPEN_BRACKETS = 'OPEN_BRACKETS'
export const CLOSE_BRACKETS = 'CLOSE_BRACKETS'
export const IS_UNIQUE = 'IS_UNIQUE'
export const SCALAR_DEFINITION = 'SCALAR_DEFINITION'
export const VARIABLE = 'VARIABLE'
export const DONE = 'DONE'
export const IN_LOOP = 'IN_LOOP'
