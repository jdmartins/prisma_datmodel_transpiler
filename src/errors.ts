export class ParseError extends Error {
  constructor(message:string='' ,...args:any) {
    super(...args)
    this.name = 'ParseError'
    this.message = message
  }
}

export class SyntaxError extends Error {
  constructor(message:string='' ,...args:any) {
    super(...args)
    this.name = 'SyntaxError'
    this.message = message
  }
}
