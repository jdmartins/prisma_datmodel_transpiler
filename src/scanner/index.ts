import tokens, { variableRegex, variableToken, Token, Tokens } from './tokens';
import { ParseError } from '../errors'

const preprocess = (dataModel: string) => {
  // TODO: Refactor this shit haha
  const extraEnters = new RegExp(/\n\n+/, 'g')
  const extraSpaces = new RegExp(/  +/, 'gm')
  const trimSpaces = new RegExp(/^ +| +$/, 'gm')
  const equality = new RegExp(/:/, 'g')
  const openBraces = new RegExp(/\[/, 'g')
  const closeBraces = new RegExp(/\]/, 'g')
  const exclamation = new RegExp(/!/, 'g')

  const withoutExtraEnters = dataModel.replace(extraEnters, '\n')
  const withEqualitySpaces = withoutExtraEnters.replace(equality, ' : ')
  const withBracesSpaces = withEqualitySpaces.replace(openBraces, ' [ ').replace(closeBraces, ' ] ')
  const withExclammationSpaces = withBracesSpaces.replace(exclamation, ' ! ')
  const withoutExtraSpaces = withExclammationSpaces.replace(extraSpaces, ' ')
  const withTrimedSpaces = withoutExtraSpaces.replace(trimSpaces, '')


  return withTrimedSpaces
}



const scan = (dataModel: string) => {
  const delimiter = new RegExp(/\s+/, 'g')
  const splittedDM:Array<string> = dataModel.split(delimiter)
  const parsedDM:Array<Token> = []

  splittedDM.forEach(word => {
    if (tokens[word as keyof Tokens] ) {
      parsedDM.push(tokens[word as keyof Tokens])
    } else if (word.match(variableRegex)) {
      parsedDM.push(variableToken(word))
    } else {
      throw new ParseError(`Unrecognized character: ${word}`)
    }
  })

  return parsedDM
}

export {
  preprocess,
  scan
}
