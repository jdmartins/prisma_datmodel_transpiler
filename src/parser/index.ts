import { Token } from '../scanner/tokens'
import { DONE, IN_LOOP } from '../constants'
import grammar from './grammar'
import { SDL, Assignment, Defintion } from './nodes'
import {
  TYPE_DEFINITION,
  ENUM_DEFINITION,
  SCALAR_DEFINITION,
  CLOSE_BRACES,
  OPEN_BRACES,
  VARIABLE,
  EQUAL
} from '../constants'

export const hasValidGrammar = (tokens: Array<Token>):boolean => {
  let context:any = grammar
  let valid:boolean = false
  let loops:number = 0
  let outsideLoop:any = {}

  tokens.forEach((each:Token) => {
    if (context[each.token] === DONE) {
      if (loops === 0) {
        valid = true
        context = grammar
      } else {
        loops = loops - 1
        context = outsideLoop
      }

    } else if (context[IN_LOOP] && context[IN_LOOP][each.token]) {
      outsideLoop = context
      loops = loops + 1
      context = context[IN_LOOP][each.token]
      valid = false

    } else if (context[each.token]) {
      context = context[each.token]
      valid = false

    } else {
      throw new Error(`Syntax error: ${each.token}`)
    }
  })

  return valid
}


export const parse = (tokens: Array<Token>):SDL => {
  const AST = new SDL()
  let actualDefinition:Defintion|null = null
  let actualAssignment:Assignment|null = null
  let insideDefinition:boolean = false
  let isAssignment:boolean = false

  tokens.forEach(each => {
    if (each.token === TYPE_DEFINITION || each.token === ENUM_DEFINITION || each.token === SCALAR_DEFINITION) {
      actualDefinition = new Defintion(each)

    } else if (each.token === CLOSE_BRACES && actualDefinition) {
      if (actualAssignment) {
        actualDefinition.addAssignment(actualAssignment)
      }

      AST.addDefinition(actualDefinition)
      insideDefinition = false

    } else if (each.token === OPEN_BRACES) {
      insideDefinition = true

    } else if (each.token === VARIABLE && actualDefinition) {
      if (insideDefinition) {
        if (actualAssignment) {
          actualDefinition.addAssignment(actualAssignment)
        }
        actualAssignment = new Assignment(each)
      } else {
        actualDefinition.setName(each)

      }

    } else if (each.token === EQUAL) {
      isAssignment = true

    } else if (isAssignment && actualAssignment) {
      isAssignment = false
      actualAssignment.setType(each)

    }
  })

  return AST
}
