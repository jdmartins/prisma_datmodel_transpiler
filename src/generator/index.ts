import { SDL, Defintion, Assignment } from '../parser/nodes'
import { capitalizeFirstLetter } from '../utils'

const defaultCode = `
scalar DateTime

scalar Long

"""An object with an ID"""
interface Node {
  """The id of the object."""
  id: ID!
}

`

export const generateCode = (AST: SDL) => {
  let code = `${defaultCode}`

  AST.typeDefinitions.forEach((definition: Defintion) => {
    code = code.concat(generateCodeForType(definition))
  })


  return code
}

const generateCodeForType = (typeDefinition: Defintion):string => {
  const assignmentsCode = typeDefinition.assignments.reduce((lastString, assignment:Assignment):string => {
    return lastString.concat(`
\t${assignment.key}: ${assignment.value}`)}, '')

  const code = `
type ${capitalizeFirstLetter(typeDefinition.name)} implements Node {${assignmentsCode}
}
  `

  return code
}
