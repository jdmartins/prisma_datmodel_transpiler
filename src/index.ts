import * as fs from 'fs'

import { preprocess, scan } from './scanner'
import { Token } from './scanner/tokens'
import { hasValidGrammar, parse } from './parser'
import { generateCode } from './generator'
import { SyntaxError } from './errors'

export const transpile = (dataModel:string):string =>{
  const tokens:Array<Token> = scan(preprocess(dataModel))

  if (!hasValidGrammar(tokens)) {
    throw new SyntaxError('Invalid Syntax')
  }

  const AST = parse(tokens)

  return generateCode(AST)
}
