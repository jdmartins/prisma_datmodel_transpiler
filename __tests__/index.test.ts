import { transpile } from '../src/'

describe('Transpiler', () => {
  it('transpiles correctly a basic datatmodel', () => {
    const dataModel:string = `type User {
      id : ID

name : String
}`

    expect(transpile(dataModel)).toMatchSnapshot()
  })
})
