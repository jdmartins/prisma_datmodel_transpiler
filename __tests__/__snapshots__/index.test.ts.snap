// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transpiler transpiles correctly a basic datatmodel 1`] = `
"
scalar DateTime

scalar Long

\\"\\"\\"An object with an ID\\"\\"\\"
interface Node {
  \\"\\"\\"The id of the object.\\"\\"\\"
  id: ID!
}


type User implements Node {
	id: ID
	name: STRING
}
  "
`;
